scanner/indexer/fetcher.go

```diff
@@ -0,0 +1,313 @@
+package indexer
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	gosync "sync"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/quay/claircore"
+	"github.com/quay/claircore/indexer"
+	"github.com/quay/zlog"
+	"github.com/stackrox/scanner/v4/internal/sync"
+	"golang.org/x/sync/errgroup"
+)
+
+var (
+	_ indexer.FetchArena = (*localFetchArena)(nil)
+	_ indexer.Realizer   = (*localFetcher)(nil)
+)
+
+// localFetchArena implements indexer.FetchArena.
+//
+// It is designed to minimize layer downloads and attempt
+// to only download a layer's contents once.
+//
+// A localFetchArena must not be copied after first use,
+// and it should be initialized via newLocalFetchArena.
+type localFetchArena struct {
```

Comment:
Would it be better if we use a name for this fetcher that exposes the fact that this fetcher minimizes layer downloads?

Also, have we checked with ClairCore to know if they are interested in this?